File: apis/csharp/common/util/RoutineInstanceBuilder.cs
using System;
using System.Collections.Generic;
using common.thrift;
using Thrift.Protocol;

namespace common.util
{
    public class RoutineInstanceBuilder
    {
        private RoutineInstanceDTO routineInstance;
        private Dictionary<String, ResourceDTO> inParams;

        public RoutineInstanceBuilder(String routineId)
        {
            routineInstance = new RoutineInstanceDTO();
            routineInstance.RoutineId = routineId;
            inParams = new Dictionary<string, ResourceDTO>();
        }

        public RoutineInstanceBuilder AddParameter<T>(String name, String dataTypeId, T value) where T : TBase
        {
            ResourceDTO resource = new ResourceDTO();
            resource.DataTypeId = dataTypeId;
            resource.Data = ThriftSerializer.Serialize(value);
            inParams.Add(name, resource);
            return this;
        }

        public RoutineInstanceBuilder AddParameter(String name, String sharedResourceId)
        {
            ResourceDTO sharedResource = new ResourceDTO();
            sharedResource.Id = sharedResourceId;
            inParams.Add(name, sharedResource);
            return this;
        }

        public RoutineInstanceDTO Build()
        {
            routineInstance.InParameters = inParams;
            return routineInstance;
        }
    }
}

File: client-api/src/main/java/at/enfilo/def/client/api/RoutineInstanceBuilder.java
package at.enfilo.def.client.api;

import at.enfilo.def.transfer.dto.ResourceDTO;
import at.enfilo.def.transfer.dto.RoutineInstanceDTO;
import org.apache.thrift.TBase;
import org.apache.thrift.TException;
import org.apache.thrift.TFieldIdEnum;
import org.apache.thrift.TSerializer;

import java.util.HashMap;
import java.util.Map;

public class RoutineInstanceBuilder {
	private static final TSerializer SERIALIZER = new TSerializer();

	private RoutineInstanceDTO routineInstance;
	private Map<String, ResourceDTO> inParams;

	public RoutineInstanceBuilder(String routineId) {
		routineInstance = new RoutineInstanceDTO();
		routineInstance.setRoutineId(routineId);
		inParams = new HashMap<>();
	}

	public RoutineInstanceBuilder addParameter(String name, TBase value) throws TException {
		ResourceDTO resource = new ResourceDTO();
		TFieldIdEnum idField = value.fieldForId((short)1); // Convention: field with id 1 is '_id'
		String dataTypeId = value.getFieldValue(idField).toString();
		resource.setDataTypeId(dataTypeId);
		resource.setData(SERIALIZER.serialize(value));
		inParams.put(name, resource);
		return this;
	}

	public RoutineInstanceBuilder addParameter(String name, String shareResourceId) {
		ResourceDTO sharedResource = new ResourceDTO();
		sharedResource.setId(shareResourceId);
		inParams.put(name, sharedResource);
		return this;
	}

	public RoutineInstanceDTO build() {
		routineInstance.setInParameters(inParams);
		return routineInstance;
	}
}
